### PR TITLE
WIP: Residual Deep NN

### PR DESCRIPTION
**Pull request by finalelement**
This PR is replacing https://github.com/dipy/dipy/pull/2096

This PR is an implementation of my own paper:

Nath, V., Schilling, K. G., Parvathaneni, P., Hansen, C. B., Hainline, A. E., Huo, Y., ... & Stepniewska, I. (2019). Deep learning reveals untapped information for local white-matter fiber reconstruction in diffusion-weighted MRI. Magnetic resonance imaging, 62, 220-227.

This is currently work in progress but I am sharing to get feedback.

### CODE REVIEW

**Comment by skoudoro** [comment]
Hi @finalelement, 

Thank you for this!

It would be good if you could present your work for the next DIPY online meeting (30 - 45 min, Next Thursday, April 23rd, 0800 PT, 1100 ET, 1600 UTC). See [this issue for more for information](https://github.com/dipy/dipy/issues/2109).  
After your talk, we can look at your PR and we can see how to help you with the integration if you have any questions.

Let us know if you are in! you can ask your question here and send me an email if you have more private questions!

**Comment by finalelement** [comment]
Hi Serge,

That sounds great. Will present on 23rd April.

Cheers,
Vish

**Comment by skoudoro** [comment]
Great! Looking forward to it! 

Thanks!

**Comment by finalelement** [comment]
@skoudoro I created my profile on figshare. After that how do I send the request, do I upload a copy of the weights and share with you? or is there something specific than I can send a request to?

**Comment by skoudoro** [comment]
1. You will have to create a fetcher. There is a nice example [here](https://github.com/dipy/dipy/blob/master/dipy/data/fetcher.py#L493). the file is `dipy.data.fetcher`. As you can see, many variables are defined: url, filenames, and the md5 value for each files
2. To generate the md5 value: you can download you file and run [get_md5(fname) function](https://github.com/dipy/dipy/blob/master/dipy/data/fetcher.py#L82).
3 when this is done, check your fetcher by calling it:
```python
from dipy.data import my_fetcher
my_fetcher()
```
4. Create a read function.

Let me know if one step is not clear.

**Comment by skoudoro** [comment]
> After that how do I send the request, do I upload a copy of the weights and share with you?

just upload your weight to get an url.

**Comment by finalelement** [comment]
@skoudoro Thanks, will let you know.

**Comment by finalelement** [comment]
@skoudoro @ShreyasFadnavis Hey guys, I got to the point of figuring out the fetcher and setting up an example script that uses the 'Stanford' HARDI data available from your prior repositories.

If you guys could take a look, try running the example script as well. It will automatically pull weights, normalize data and make Histology based fODF predictions. Also tested a small visualization of it.

**Comment by finalelement** [comment]
Hey guys, hope y'all find some time to run this code before the upcoming meeting. @skoudoro @ShreyasFadnavis .. or If I you think I am tagging the wrong people, let me know who should be. :)

Cheers,
Vish

**Comment by skoudoro** [comment]
Thank you for pinging me @finalelement. I will look into it today or tomorrow. Sorry for the delay.

**Comment by ShreyasFadnavis** [comment]
> Hey guys, hope y'all find some time to run this code before the upcoming meeting. @skoudoro @ShreyasFadnavis .. or If I you think I am tagging the wrong people, let me know who should be. :)
> 
> Cheers,
> Vish

Sorry for the slowness in my response! Will try it out 👍 
I intend to start out with the example... Any heads up for me?

**Comment by finalelement** [comment]
Well the only heads up I have is, I need suggestions on improving the code
so that it matches up to how the rest of dipy stuff has been written.
Otherwise it should be fairly simple to run it :)

Thanks
Vish

**Comment by finalelement** [comment]
@skoudoro @ShreyasFadnavis Any updates on this specific PR from your end?

**Comment by frheault** [comment]
@finalelement Bennett just warn me to look for this PR, I remember my comments from the previous PR which was closed so I never saw this new PR !

I will try it this week. Do you think its possible to update with master and let the tests re-run?

File: dipy/nn/resdnn.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from dipy.reconst.shm import sf_to_sh, just_sh_basis
+from dipy.nn.resdnn_histology import Histo_ResDNN
+from dipy.core.sphere import Sphere, HemiSphere
+from dipy.data import default_sphere
+from dipy.viz import window, actor
+
+def train():
+
+    # Load The Gradient Directions and form a Sphere
+    grad_mat_path = r'/nfs/masi/nathv/dipy_data_2020/monkey_stuff/ng_100.mat'
```
Comment (frheault) [inline dipy/nn/resdnn.py:40]: This should be handle by a fetcher or at least an URL

File: dipy/nn/resdnn_hcp_example.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import os
```
Comment (frheault) [inline dipy/nn/resdnn_hcp_example.py:1]: Move to https://github.com/dipy/dipy/tree/master/doc/examples

File: dipy/nn/resdnn_hcp_example.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+from dipy.core.sphere import Sphere, HemiSphere
+from dipy.data import default_sphere
+from dipy.viz import window, actor
+
+
+# Build the ResDNN Network & Load Weights
+# TODO The Hard-coded path needs a fetcher here
+resdnn_model = Histo_ResDNN()
+#print(resdnn_model.model.weights)
+model_weights_path = r'/nfs/masi/nathv/dipy_data_2020/monkey_stuff/monkey_weights_resdnn_mri_2018.h5'
```
Comment (frheault) [inline dipy/nn/resdnn_hcp_example.py:19]: Fetcher

File: dipy/nn/resdnn_histology_example.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+mean_b0 = data[:, :, :, b0_indices]
+mean_b0 = np.squeeze(np.nanmean(mean_b0, axis=4))
+print(mean_b0.shape)
+
+# Detect number of b-values and extract a single shell of DW-MRI Data
+unique_shells = np.unique(bvals)
+print('Number of b-values: {}'.format(np.unique(bvals)))
+
+# Going with the assumption that the first shell would be b0's
```
Comment (frheault) [inline dipy/nn/resdnn_histology_example.py:34]: If this is an example, phrase this as an instruction so people can generalize to their data

File: dipy/reconst/shm.py [hunk 5]
```diff
@@ -911,6 +911,43 @@ def sf_to_sh(sf, sphere, sh_order=4, basis_type=None, smooth=0.0):
 
     return sh
 
+def just_sh_basis(sphere, sh_order=4, basis_type=None):
```
Comment (frheault) [inline dipy/reconst/shm.py:914]: Use sh_to_sf_matrix (from the same shm.py) instead

**Review by frheault — COMMENTED** [review]
Could you remove every file path that points to a specific computer.
Also it would help to remove all comments that are old irrelevant code. It confuses me to see debugging code in a PR

**Comment by Garyfallidis** [comment]
Ping @finalelement

**Comment by skoudoro** [comment]
closing in favor of #2520

**Pull request closed without merging**